### ISSUE TITLE
32-bit Coq 8.8.0 and 8.7.2 build failure under switch 4.06.1+32bit from Opam on OSX

### ISSUE THREAD

**Issue by sumahadevan** [coq/coq#7480]
#### Version

BOTH Coq 8.8.0 and 8.7.2 under OCaml switch 4.06.1+32bit

#### Operating system

macOS High Sierra 10.13.4 = Darwin localhost 17.5.0 Darwin Kernel Version 17.5.0: Fri Apr 13 19:32:32 PDT 2018; root:xnu-4570.51.2~1/RELEASE_X86_64 x86_64

#### Description of the problem

The OCaml compiler that gives the Coq build error is 4.06.1+32bit. It is a new switch and NOT the sytem compiler that previously got installed by Homebrew on Mac. This new compiler was installed by successfully running

https://github.com/ejgallego/jscoq/blob/v8.8/toolchain-setup.sh

as that project needs a 32-bit Coq build for js_of_ocaml (the 'make coq' step failed for the above repository as well, after which I tried it directly through OPAM with the same results given below).

I was successful previously in setting up a new 64-bit OCaml 4.06.1 switch, installing Coq 8.7.2 via Opam, and building two Coq libraries (math-classes and corn) from the coq-released repository (both via Opam and directly), so my setup is working fine otherwise (at least with 64-bit OCaml 4.06.1).

The error is similar for Coq 8.8.0 and 8.7.2 but the error logs are slightly different, hence am reproducing both below.

opam install coq.8.8.0

```
The following actions will be performed:
  ∗  install coq 8.8.0

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[default] https://opam.ocaml.org/archives/coq.8.8.0+opam.tar.gz downloaded

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[ERROR] The compilation of coq failed at "make -j4".
Processing  1/1: [coq: rm]
#=== ERROR while installing coq.8.8.0 =========================================#
# opam-version 1.2.2
# os           darwin
# command      make -j4
# path         /Users/mads/.opam/4.06.1+32bit/build/coq.8.8.0
# compiler     4.06.1+32bit
# exit-code    2
# env-file     /Users/mads/.opam/4.06.1+32bit/build/coq.8.8.0/coq-8836-d2d111.env
# stdout-file  /Users/mads/.opam/4.06.1+32bit/build/coq.8.8.0/coq-8836-d2d111.out
# stderr-file  /Users/mads/.opam/4.06.1+32bit/build/coq.8.8.0/coq-8836-d2d111.err
### stdout ###
# [...]
# OCAMLOPT  plugins/ltac/evar_tactics.ml
# OCAMLOPT  plugins/ltac/tactic_option.ml
# OCAMLOPT  plugins/ltac/extraargs.ml
# OCAMLOPT  plugins/ltac/g_auto.ml
# OCAMLOPT  plugins/ltac/g_class.ml
# OCAMLOPT  plugins/ltac/rewrite.ml
# OCAMLOPT  plugins/ltac/g_ltac.ml
# OCAMLOPT -a -o ide/coqidetop.cmxa
# COQMKTOP -o bin/coqtop
# OCAMLOPT  plugins/ltac/g_obligations.ml
### stderr ###
# [...]
# File "_none_", line 1:
# Warning 58: no cmx file was found in path for module Dynlink, and its interface was not compiled with -opaque
# File "toplevel/coqtop_opt_bin.ml", line 1:
# Error: No implementations provided for the following modules:
#          Dynlink referenced from kernel/kernel.cmxa(Nativelib),
#            library/library.cmxa(Library), vernac/vernac.cmxa(ExplainErr),
#            vernac/vernac.cmxa(Mltop)
# make[1]: *** [bin/coqtop] Error 2
# make[1]: *** Waiting for unfinished jobs....
# make: *** [submake] Error 2



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
The following actions failed
  ∗  install coq 8.8.0
No changes have been performed
```

The log for Coq 8.7.2 looks slightly different.

opam install coq.8.7.2

```
The following actions will be performed:
  ∗  install coq 8.7.2

=-=- Gathering sources =-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[default] https://opam.ocaml.org/archives/coq.8.7.2+opam.tar.gz downloaded

=-=- Processing actions -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
[ERROR] The compilation of coq failed at "make -j4".
Processing  1/1: [coq: rm]
#=== ERROR while installing coq.8.7.2 =========================================#
# opam-version 1.2.2
# os           darwin
# command      make -j4
# path         /Users/mads/.opam/4.06.1+32bit/build/coq.8.7.2
# compiler     4.06.1+32bit
# exit-code    2
# env-file     /Users/mads/.opam/4.06.1+32bit/build/coq.8.7.2/coq-13883-d2d111.env
# stdout-file  /Users/mads/.opam/4.06.1+32bit/build/coq.8.7.2/coq-13883-d2d111.out
# stderr-file  /Users/mads/.opam/4.06.1+32bit/build/coq.8.7.2/coq-13883-d2d111.err
### stdout ###
# [...]
# OCAMLC    pretyping/locusops.mli
# OCAMLC    interp/smartlocate.mli
# OCAMLC    proofs/miscprint.mli
# OCAMLC    tactics/term_dnet.mli
# OCAMLC    plugins/extraction/miniml.mli
# OCAMLOPT  plugins/nsatz/ideal.ml
# OCAMLOPT  checker/univ.ml
# OCAMLBEST -o plugins/micromega/csdpcert
# OCAMLOPT  tools/fake_ide.ml
# OCAMLBEST -o bin/coqdep
### stderr ###
# [...]
# Redundant [TYPED AS] clause in [ARGUMENT EXTEND ssrindex].
# clang: warning: -Wl,-read_only_relocs,suppress: 'linker' input unused [-Wunused-command-line-argument]
# clang: clang: warning: warning: -Wl,-read_only_relocs,suppress: 'linker' input unused [-Wunused-command-line-argument]
# -Wl,-read_only_relocs,suppress: 'linker' input unused [-Wunused-command-line-argument]
# clang: warning: -Wl,-read_only_relocs,suppress: 'linker' input unused [-Wunused-command-line-argument]
# File "_none_", line 1:
# Error: Cannot find file dynlink.cmxa
# make[1]: *** [bin/coqdep] Error 2
# make[1]: *** Waiting for unfinished jobs....
# make: *** [submake] Error 2



=-=- Error report -=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=  🐫 
The following actions failed
  ∗  install coq 8.7.2
No changes have been performed
```

Although I found several references to 32-bit OCaml build issues with dynlink on Mac (for example for camlp4), I am not sure if the above build issue for Coq is a known issue.

**Comment by ejgallego** [comment]
I think this is just a duplicate of 6658 but I'll have to check.

**Comment by ejgallego** [comment]
Hard for me to debug due to the platform.

**Comment by herbelin** [comment]
Looks indeed like #6658. I confirm that `dynlink.cmx` is not produced by `opam install 4.06.1+32bit`.

@sumahadevan: Can you check that the recipe mentioned in #6658 works for you?

**Comment by sumahadevan** [comment]
@herbelin, I will be happy to try but may need some inputs. I looked at the recipe mentioned in #6658, where you have indicated that

> I could reproduce your issue which is actually a known one. You shall find the fix in the header of file dev/dynlink.ml. The header is not informative enough: you should also install dynlink.a and overwrite dynlink.cmi. Then, I could make Coq 8.7 compilable with 4.06 32 bits on OS X, using option -natdynlink no.

I located the file `dev/dynlink.ml` and went through the instructions in its header, which I reproduce below. Can you explicitly indicated what additional commands are to be executed in addition to the ones given there?

```
(** Some architectures may have a native ocaml compiler but no native
    dynlink.cmxa (e.g. ARM). If you still want to build a native coqtop
    there, you'll need this dummy implementation of Dynlink.
    Compile it and install with:

      ocamlopt -a -o dynlink.cmxa dynlink.ml
      sudo cp -i dynlink.cmxa `ocamlopt -where`

    Then build coq this way: ./configure -natdynlink no && make world
*)
```

Also, is the `sudo` really necessary (unless the `cp -i` places `dynlink.cmxa` in a system directory, I couldn't quite understand what that step did)? All the 4.06.1+32bit compiler files are in an OPAM switch under my home folder.

I would be grateful if you could provide the additional commands to be executed for the steps you mentioned for `dynlink.a` and `dynlink.cmi` for my better understanding. Of course, I can see from the logs of the manual Coq compilation in https://github.com/ejgallego/jscoq/issues/29 that right now by default on OSX we have `Native dynamic link support : true`.

**Comment by sumahadevan** [comment]
I was able to clarify the query about the `sudo`, it is not required in my case. A little googling told me that `ocamlopt -where` returns the location of the OCaml standard library. Executing it in my case gives `/Users/mads/.opam/4.06.1+32bit/lib/ocaml` so at least now I have a concrete location for where the `dynlink` files are supposed to go (makes things somewhat easier!).

**Comment by sumahadevan** [comment]
So I have built `dynlink.cmxa` and copied it as well as `dynlink.a` and `dynlink.cmi` to my `ocamlopt -where` location (renamed the original  `dynlink.cmi` there). @herbelin, no need for details of the copy commands (glad I could figure it out). I will now try to build Coq 8.8 as indicated.

**Comment by sumahadevan** [comment]
@herbelin, I tried to build both Coq 8.8.0 and Coq 8.7.2. The good news is that the recipe works for Coq 8.7.2. Unfortunately it does not work for Coq 8.8.0. This may be related to the different way in which the error manifests itself in the original OPAM logs above (before the recipe). The Coq build logs after applying the recipe are attached.

[coq-8.8-build-ocaml-4.06.1-32bit-after-dynlink-patch-osx.txt](https://github.com/coq/coq/files/1997309/coq-8.8-build-ocaml-4.06.1-32bit-after-dynlink-patch-osx.txt)
[coq-8.7-build-ocaml-4.06.1-32bit-after-dynlink-patch-osx.txt](https://github.com/coq/coq/files/1997310/coq-8.7-build-ocaml-4.06.1-32bit-after-dynlink-patch-osx.txt)

Essentially in 8.7.2 the original error was 

```
# Error: Cannot find file dynlink.cmxa
# make[1]: *** [bin/coqdep] Error 2
```

which is now resolved by the recipe. In 8.8.0 the original error was

```
# File "toplevel/coqtop_opt_bin.ml", line 1:
# Error: No implementations provided for the following modules:
#          Dynlink referenced from kernel/kernel.cmxa(Nativelib),
#            library/library.cmxa(Library), vernac/vernac.cmxa(ExplainErr),
#            vernac/vernac.cmxa(Mltop)
# make[1]: *** [bin/coqtop] Error 2
```
and this error persists unchanged even after the fix made in the recipe.

I had essentially identical behavior building Coq 8.8.0 and 8.7.2, after applying the recipe, for the same issue in ejgallego/jscoq#29, namely 8.8.0 failed at the same point while 8.7.2 went ahead successfully (though there was a separate issue later on in the 8.7.2 `make coq` there, which I will update shortly).

**Comment by herbelin** [comment]
To @ejgallego: Technically speaking, the difference of behavior between 8.7.2 and 8.8.0 comes from the way ocaml libraries are found (598e3ae4). In 8.7.2, the list of libraries was explicitly given (namely `ocamlfind opt ... str.cmxa unix.cmxa nums.cmxa dynlink.cmxa threads.cmxa gramlib.cmxa ...`). In 8.8.0, the list of libraries is to be found by `ocamlfind` (namely using `ocamlfind opt ... -package num,str,unix,dynlink,threads -package camlp5.gramlib ...`).

The reason `ocamlfind` does not find `~/.opam/4.06.1+32bit/lib/ocaml/dynlink.cmxa` seems related to what is in the file `~/.opam/4.06.1+32bit/lib/dynlink/META`: `dynlink.cmxa` is missing there.

To @sumahadevan: The following extended recipe eventually worked for me:

- edit `~/.opam/4.06.1+32bit/lib/dynlink/META` to replace the line `archive(native) = ""` by `archive(native) = "dynlink.cmxa"`
- in `dev`: `ocamlopt -a -o dynlink.cmxa dynlink.ml`
- in `dev`: ``cp -i dynlink.cmxa dynlink.cmi dynlink.a ~/.opam/4.06.1+32bit/lib/ocaml``
- at the root of Coq archive: `./configure -natdynlink no && make alienclean && make -j4`

I don't know if manually editing `~/.opam/4.06.1+32bit/lib/dynlink/META` and overriding `~/.opam/4.06.1+32bit/lib/ocaml/dynlink.cmi` is a clean way to proceed but it worked.

**Comment by ejgallego** [comment]
Hi folks, sorry for the problems with this, let's see how we can fix that better. A couple of thoughts:

- if this is a duplicate of #6658 I suggest we close this issue and continue the discussion there.
- main point, what's the upstream [OCaml] position about 32bit dynlinking in OSX? Is the fact that 32bit dynlink doesn't work there a problem of the OPAM package or just a limitation of OCaml itself? Is there not a Mantis issue?

If it turns out that we could fix OCaml to work with native dynlink that would IMO lower the priority of this bug.

Regarding a workaround, I am indeed not sure what is the best option here. I guess that if no native dynlink is found then the best would be just not to build the native toolchain at all, right?

Coqtop is not going to be very useful if we cannot load plugins, I'm afraid.

**Comment by ejgallego** [comment]
@herbelin solution is correct, the problem comes from the META file, it would be indeed possible to workaround this automatically by shipping in Coq a `dynlink.fake` package, but even better would be to actually detect Dynlink.

**Comment by ejgallego** [comment]
Some more information on the upstream issue: https://github.com/ocaml/ocaml/pull/692

**Comment by sumahadevan** [comment]
I had a question about building `coqtop.byte`, which the `make coq` for downstream project https://github.com/ejgallego/jscoq also builds. There `coqtop` built successfully after the initial recipe was applied, but later in the same build `coqtop.byte` gave the followng error

```
COQMKTOP -o bin/coqtop.byte
findlib: [WARNING] Interface profile.cmi occurs in several directories: /Users/mads/.opam/4.06.1+32bit/lib/ocaml/compiler-libs, lib
findlib: [WARNING] Interface topdirs.cmi occurs in several directories: /Users/mads/.opam/4.06.1+32bit/lib/ocaml/compiler-libs, /Users/mads/.opam/4.06.1+32bit/lib/ocaml
File "/var/folders/pl/9pp9lqm95p76pjrj3t5wgr9w0000gn/T/coqmain654b0f.ml", line 1:
Error: Files kernel/kernel.cma(Nativelib)
       and /Users/mads/.opam/4.06.1+32bit/lib/ocaml/dynlink.cma(Dynlink)
       make inconsistent assumptions over interface Dynlink
make[2]: *** [bin/coqtop.byte] Error 2
```

I will try to replicate this by building Coq 8.7.2 directly with the same options used in that project and let you know. I am not sure what is the impact of our recipe on the build settings for the bytecode compiler, so thought I should bring this to your attention (and will update you after building it directly with same settings).

**Comment by herbelin** [comment]
I think I see. The `dynlink.cmi` file is modified, so the `dynlink.cma` file needs to be modified too. Try completing with the following ingredients
- edit `~/.opam/4.06.1+32bit/lib/dynlink/META` to replace the line `archive(native) = ""` by `archive(native) = "dynlink.cmxa"`
- in `dev`: `ocamlopt -a -o dynlink.cmxa dynlink.ml`
- in `dev`: `ocamlc -a -o dynlink.cma dynlink.ml`
- in `dev`: ``cp -i dynlink.cmxa dynlink.cma dynlink.cmi dynlink.a ~/.opam/4.06.1+32bit/lib/ocaml``
- at the root of Coq archive: `./configure -natdynlink no && make alienclean && make -j4`

Then, hopefully, `bin/coqtop.byte` will work.

PS to dev: I don't know why but I could not build from the `dynlink.cmi` in `~/.opam/4.06.1+32bit/lib/ocaml` and I had to redo it.

**Comment by SkySkimmer** [comment]
Why are we putting the cma in ocaml's directory, couldn't we put it in our own lib/ or some such?

**Comment by herbelin** [comment]
> Why are we putting the cma in ocaml's directory, couldn't we put it in our own lib/ or some such?

Maybe.

I guess that at the origin, Pierre's idea was to do as if the files had been provided by ocaml.

If in a local `lib`, wouldn't we have to use different makefile targets depending on whether `dynlink` is provided or not, since, currently, it is supposed to be found by `ocamlfind`.

PS: I'm actually not an expert on this. I can make some tests as a OS X user if that helps but I could hardly spend more time on it.

**Comment by ejgallego** [comment]
Yup I think the best strategy is to provide a fake local dynlink that is used as a fallback.

But again, if we have no native code dynlink, shouldn't we just refuse to build any opt targets?

**Comment by sumahadevan** [comment]
Everything worked beautifully with the final recipe, for BOTH native and bytecode compilation in Coq 8.8.0 and 8.7.2 versions, whether compiling Coq directly or using `make coq` in the downstream project https://github.com/ejgallego/jscoq. In more detail:

* I first applied the latest recipe above from @herbelin - it should be noted that these compiled `dynlink` files are from Coq 8.8.
* Without editing the dynlink `META` file yet (as it is not required in the case of 8.7.2), I confirmed that the 8.7.2  `coqtop.byte` also builds successfully now (using the downstream project `make coq` 8.7 version).
* I now edited the dynlink `META` file as per the final recipe (I noticed that this file already had a valid entry for `archive(byte)`, it was only the entry for `archive(native)` that was missing).
* I cloned the 8.8 branch of the Coq repository and successfully built both the native and bytecode versions (adding `make byte` to the final recipe above). It should be noted that I did NOT recompile the dynlink files after editing the `META` file, because the `META` file is not relevant for their compilation, but only for `ocamlfind` at run-time. 
* For abundant caution, I cloned the 8.7 branch of the Coq repository and again built both the native and bytecode versions successfully (only difference here is that `make alienclean` is not part of Coq 8.7 branch).
* @ejgallego, finally I successfully built the Coq 8.8 native and bytecode versions using  `make coq` from the above downstream repository.

I found the recipe to be fairly straightforward to apply (it needed to be applied only once), and of course its application is eventually likely to be automated in some way.

**Comment by sumahadevan** [comment]
I think I was a little confused about the architecture check. I realized that this checks the intrinsic architecture of the machine, and on all recent Intel-based Macs it will return `true`. However, it should be possible to detect that the OCaml compiler being used is 32-bit (regardless of the architecture being 64-bit) and dynlink accordingly?

**Comment by ejgallego** [comment]
Is this fixed?

**Comment by ejgallego** [comment]
It seems that support for 32bit binaries is going to be removed from OSX very soon, if I understand correctly. https://support.apple.com/en-us/HT208436

Thus we cannot do a lot on this front.
